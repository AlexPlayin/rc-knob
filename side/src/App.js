import React from 'react'

import { H1, H2, Grid, Description } from './styled'
import PointerExample from './examples/Pointer'
import ScaleExample from './examples/Scale'
import ArcExample from './examples/Arc'
import ArcPointerExample from './examples/Arc_Pointer'
import ScalePointerExample from './examples/Scale_Pointer'
import ValueExample from './examples/Value'
import ArcPointerValueExample from './examples/Arc_Pointer_Value'
import ScaleScale from './examples/Scale_Scale'
import PointerCustom from './examples/Pointer_custom'
import ScaleCustom from './examples/Scale_custom'
import ScaleStyled from './examples/Scale_styled'
import ScaleStyled2 from './examples/Scale_styled_2'
import ArcPointerStyled from './examples/Arc_Pointer_styled'
import EventMouse from "./examples/Event_mouse";
import EventReadOnly from "./examples/Event_readOnly";
import EventNoWheel from "./examples/Event_nowheel";


import Filter from './filter'

export default () => {
    return (
        <div>
            <Filter />
            <H1>React Knob</H1>
            <H2>
                <span>Base Components</span>
            </H2>
            <Description>
                The base of all knobs is the <pre>Knob</pre> component which handles
                all of the user interaction and calculation of the knob value. The
                user can interact via drag'n drop, mousewheel and keyboard. The
                knobs are accessible by the <pre>tab</pre> key.
                <br /> The display part is splitted into the <pre>
                    Pointer
                </pre>, <pre>Scale</pre>, <pre>Arc</pre> and <pre>Value</pre>. One
                or more of them can be <pre>children</pre> of a <pre>Knob</pre>.
            </Description>
            <Grid>
                <PointerExample />
                <ArcExample />
                <ScaleExample />
                <ValueExample />
            </Grid>
            <H2>
                <span>Combinations</span>
            </H2>
            <Description>
                You can combine all of the base components in a knob.
            </Description>
            <Grid>
                <ArcPointerExample />
                <ScalePointerExample />
                <ArcPointerValueExample />
                <ScaleScale />
            </Grid>
            <H2>
                <span>Custom Elements</span>
            </H2>
            <Description>
                The <pre>Pointer</pre> and <pre>Scale</pre> components can also
                render custom components.
            </Description>
            <Grid>
                <PointerCustom />
                <ScaleCustom />
            </Grid>
            <H2>
                <span>Events</span>
            </H2>
            <Grid>
                <EventMouse />
                <EventReadOnly />
                <EventNoWheel />
            </Grid>
            <H2>
                <span>Styles</span>
            </H2>
            <Description>
                The components and the knob itself can be styled by addingthe{' '}
                <pre>className</pre> prop. The <pre>Scale</pre> can have an
                additional <pre>activeClassname</pre> prop to style the active tick
                differently. As all components are SVG you can also use filters in
                your styles.
            </Description>
            <Grid>
                <ArcPointerStyled />
                <ScaleStyled />
                <ScaleStyled2 />
            </Grid>
        </div>
    )
}
