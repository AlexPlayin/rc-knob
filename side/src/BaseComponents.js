import React from 'react'

import { H1, H2, Grid, Description } from './styled'
import PointerExample from './examples/Pointer'
import ScaleExample from './examples/Scale'
import ArcExample from './examples/Arc'
import ArcPointerExample from './examples/Arc_Pointer'
import ScalePointerExample from './examples/Scale_Pointer'
import ValueExample from './examples/Value'
import ArcPointerValueExample from './examples/Arc_Pointer_Value'
import ScaleScale from './examples/Scale_Scale'
import PointerCustom from './examples/Pointer_custom'
import ScaleCustom from './examples/Scale_custom'

export default () => (
    <div>
        <H1>React Knob</H1>
        <H2>
            <span>Base Components</span>
        </H2>
        <Description>
            The base of all knobs is the <pre>Knob</pre> component which handles
            all of the user interaction and calculation of the knob value.{' '}
            <br /> The display part is splitted into the <pre>Pointer</pre>,{' '}
            <pre>Scale</pre>, <pre>Arc</pre> and <pre>Value</pre>. One or more{' '}
            of them can be <pre>children</pre> of a <pre>Knob</pre>.
        </Description>
        <Grid>
            <PointerExample />
            <ArcExample />
            <ScaleExample />
            <ValueExample />
        </Grid>
        <H2>
            <span>Combinations</span>
        </H2>
        <Description>
            You can combine all of the base components in a knob.
        </Description>
        <Grid>
            <ArcPointerExample />
            <ScalePointerExample />
            <ArcPointerValueExample />
            <ScaleScale />
        </Grid>
        <H2>
            <span>Custom Elements</span>
        </H2>
        <Description>
            The <pre>Knob</pre> component and the <pre>Pointer</pre> and{' '}
            <pre>Scale</pre> components can also render custom components. All
            direct children of the <pre>Knob</pre> get the following props
            passed: <pre>percentage</pre> <pre>size</pre>, <pre>value</pre>,
            <pre>angleOffset</pre>,<pre>angleRange</pre>,{' '}
            <pre>radius: size / 2</pre>,<pre>center: size / 2</pre>,{' '}
            <pre>steps</pre>. The children of <pre>Pointer</pre> get{' '}
            <pre>width</pre>, <pre>height</pre>, <pre>percentage</pre> passed.
        </Description>
        <Grid>
            <PointerCustom />
            <ScaleCustom />
        </Grid>
    </div>
)
