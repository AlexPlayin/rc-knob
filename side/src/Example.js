import React from 'react'
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter'
import { H3, Cell, CodeBlock } from './styled'
import { Knob } from 'rc-knob'
import theme from './prismTheme'

const renderKnob = (child, options = {}) => (
    <Knob
        onChange={value => console.log(value)}
        size={100}
        value={0}
        angleOffset={220}
        angleRange={280}
        min={0}
        max={100}
        {...options}
        snap
    >
        {child}
    </Knob>
)

export default ({ example, knob, title, options }) => (
    <Cell>
        <H3>{title}</H3>
        {renderKnob(knob, options)}
        <CodeBlock>
            <SyntaxHighlighter language="jsx" style={theme}>
                {example}
            </SyntaxHighlighter>
        </CodeBlock>
    </Cell>
)
