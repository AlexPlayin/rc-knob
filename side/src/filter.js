import React from 'react'

export default () => (
    <svg>
        <filter id="inner-shadow">
            <feOffset dx="0" dy="5" />
            <feGaussianBlur stdDeviation="5" result="offset-blur" />
            <feComposite
                operator="out"
                in="SourceGraphic"
                in2="offset-blur"
                result="inverse"
            />
        </filter>
    </svg>
)
