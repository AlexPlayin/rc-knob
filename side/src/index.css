@import url('https://fonts.googleapis.com/css?family=Work+Sans:100,300');
@import url('https://fonts.googleapis.com/css?family=Eczar:800');

html {
    height: 100%;
}

body {
    margin: 0;
    padding: 0;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto',
        'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
        'Helvetica Neue', sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    height: 100%;
    padding-top: 254px;
}

code {
    font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
        monospace;
}
.value {
    font-family: 'Eczar', serif;
    fill: #fc5a96;
    font-size: 50px;
}
.value2 {
    font-family: 'Work Sans', sans-serif;
    fill: #fc5a96;
    font-size: 30px;
}
