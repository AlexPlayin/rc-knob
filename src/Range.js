import React from 'react'

const pointOnCircle = (center, radius, angle) => ({
    x: center + radius * Math.cos(angle),
    y: center + radius * Math.sin(angle),
})
const degTorad = deg => (Math.PI * deg) / 180

const calcPath = ({
    percentageFrom,
    percentageTo,
    angleOffset,
    angleRange,
    arcWidth,
    radius: outerRadius,
    center,
}) => {
    const angle = angleRange * (percentageTo - percentageFrom)
    const angleFrom = angleOffset - 90 + angleRange * percentageFrom
    const innerRadius = outerRadius - arcWidth
    const angleFromRad = degTorad(angleFrom)
    const angleToRad = degTorad(angleFrom + angle)
    const largeArcFlag = Math.abs(angle) < 180 ? 0 : 1
    const direction = angle >= 0 ? 1 : 0

    const p1 = pointOnCircle(center, outerRadius, angleFromRad)
    const p2 = pointOnCircle(center, outerRadius, angleToRad)
    const p3 = pointOnCircle(center, innerRadius, angleToRad)
    const p4 = pointOnCircle(center, innerRadius, angleFromRad)

    return `M${p1.x},${
        p1.y
    } A${outerRadius},${outerRadius} 0 ${largeArcFlag} ${direction} ${p2.x},${p2.y}L${
        p3.x
    },${p3.y} A${innerRadius},${innerRadius} 0 ${largeArcFlag} ${1-direction} ${p4.x},${
        p4.y
    } L${p1.x},${p1.y}`
}

export const Range = ({
    color,
    percentage=null,
    percentageFrom=null,
    percentageTo=null,
    ...props
}) => {
    let pfrom, pto
    if (percentageFrom !== null && percentageTo !== null) {
        pfrom = percentageFrom
        pto = percentageTo
    } else if (percentageFrom !== null) {
        pfrom = percentageFrom
        pto = percentage
    } else if (percentageTo !== null) {
        pfrom = percentage
        pto = percentageTo
    } else {
        pfrom = 0
        pto = percentage
    }
    // Clamp
    if (Math.abs(pto - pfrom) >= 1) {
        pfrom = 0
        pto = 0.9999
    }

    const d = calcPath({percentageFrom:pfrom, percentageTo:pto, ...props})
    return (<g>
        <path d={d} style={{ fill: color }} />
    </g>)
}
