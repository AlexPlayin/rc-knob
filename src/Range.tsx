import React from 'react';
import type { KnobState, PropsWithKnobState } from 'types';

const pointOnCircle = (center: number, radius: number, angle: number) => ({
    x: center + radius * Math.cos(angle),
    y: center + radius * Math.sin(angle),
});
const degTorad = (deg: number) => (Math.PI * deg) / 180;

const clampDeg = (deg: number) =>
    deg >= 360 ? 359.999 : deg <= -360 ? -359.999 : deg;

const calcPath = ({
    percentageFrom,
    percentageTo,
    angleOffset,
    angleRange,
    arcWidth,
    radius: outerRadius,
    center,
}: {
    percentageFrom: number;
    percentageTo: number;
    angleOffset: number;
    angleRange: number;
    arcWidth: number;
    radius: number;
    center: number;
}) => {
    const angle = angleRange * (percentageTo - percentageFrom);
    const clampedAngle = clampDeg(angle);
    const angleFrom = angleOffset - 90 + angleRange * percentageFrom;
    const innerRadius = outerRadius - arcWidth;
    const angleFromRad = degTorad(angleFrom);
    const angleToRad = degTorad(angleFrom + clampedAngle);
    const largeArcFlag = Math.abs(clampedAngle) < 180 ? 0 : 1;
    const direction = clampedAngle >= 0 ? 1 : 0;

    const p1 = pointOnCircle(center, outerRadius, angleFromRad);
    const p2 = pointOnCircle(center, outerRadius, angleToRad);
    const p3 = pointOnCircle(center, innerRadius, angleToRad);
    const p4 = pointOnCircle(center, innerRadius, angleFromRad);

    return `M${p1.x},${
        p1.y
    } A${outerRadius},${outerRadius} 0 ${largeArcFlag} ${direction} ${p2.x},${
        p2.y
    }L${p3.x},${p3.y} A${innerRadius},${innerRadius} 0 ${largeArcFlag} ${
        1 - direction
    } ${p4.x},${p4.y} L${p1.x},${p1.y}`;
};

interface Props {
    color: string;
    arcWidth: number;
    percentageFrom: number | null;
    percentageTo: number | null;
    radius?: number;
    outerRadius?: number;
}

export const Range = ({
    color,
    percentage,
    percentageFrom = null,
    percentageTo = null,
    ...props
}: PropsWithKnobState<Props>) => {
    let pfrom: number | null;
    let pto: number | null;
    if (percentageFrom !== null && percentageTo !== null) {
        pfrom = percentageFrom;
        pto = percentageTo;
    } else if (percentageFrom !== null) {
        pfrom = percentageFrom;
        pto = percentage;
    } else if (percentageTo !== null) {
        pfrom = percentage;
        pto = percentageTo;
    } else {
        pfrom = 0;
        pto = percentage;
    }
    if (pfrom === null || pto === null) {
        return <></>;
    }
    const d = calcPath({ ...props, percentageFrom: pfrom, percentageTo: pto });
    return (
        <g>
            <path d={d} style={{ fill: color }} />
        </g>
    );
};
