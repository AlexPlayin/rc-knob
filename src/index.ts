export * from './Knob'
export * from './Arc'
export * from './Scale'
export * from './Pointer'
export * from './Value'
export * from './Range'
export * from './Spiral'
export * from './Label'
export * from './types'
