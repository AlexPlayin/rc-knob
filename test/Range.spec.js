import React from 'react'
import { shallow } from 'enzyme'
import { Range } from '../src/Range'

describe('Arc', () => {
    it('renders correct clockwise small', () => {
        const component = shallow(
            <Range
                percentageFrom={0.2}
                percentageTo={0.5}
                angleOffset={0}
                angleRange={360}
                arcWidth={5}
                center={0}
                radius={50}
                color="lime"
                background="red"
            />
        )
        expect(component).toMatchSnapshot()
    })
    it('renders correct clockwise large', () => {
        const component = shallow(
            <Range
                percentageFrom={0.2}
                percentageTo={0.8}
                angleOffset={0}
                angleRange={360}
                arcWidth={5}
                center={0}
                radius={50}
                color="lime"
                background="red"
            />
        )
        expect(component).toMatchSnapshot()
    })
    it('renders correct anticlockwise small', () => {
        const component = shallow(
            <Range
                percentageFrom={0.2}
                percentageTo={0.5}
                angleOffset={0}
                angleRange={-360}
                arcWidth={5}
                center={0}
                radius={50}
                color="lime"
                background="red"
            />
        )
        expect(component).toMatchSnapshot()
    })
    it('renders correct anticlockwise large', () => {
        const component = shallow(
            <Range
                percentageFrom={0.2}
                percentageTo={0.8}
                angleOffset={0}
                angleRange={-360}
                arcWidth={5}
                center={0}
                radius={50}
                color="lime"
                background="red"
            />
        )
        expect(component).toMatchSnapshot()
    })
})
