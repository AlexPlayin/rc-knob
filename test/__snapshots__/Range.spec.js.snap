// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arc renders correct anticlockwise large 1`] = `
<g>
  <path
    d="M-47.552825814757675,-15.450849718747376 A50,50 0 1 0 47.55282581475768,-15.450849718747358L42.797543233281914,-13.905764746872622 A45,45 0 1 1 -42.79754323328191,-13.905764746872638 L-47.552825814757675,-15.450849718747376"
    style={
      Object {
        "fill": "lime",
      }
    }
  />
</g>
`;

exports[`Arc renders correct anticlockwise small 1`] = `
<g>
  <path
    d="M-47.552825814757675,-15.450849718747376 A50,50 0 0 0 -9.184850993605149e-15,50L-8.266365894244634e-15,45 A45,45 0 0 1 -42.79754323328191,-13.905764746872638 L-47.552825814757675,-15.450849718747376"
    style={
      Object {
        "fill": "lime",
      }
    }
  />
</g>
`;

exports[`Arc renders correct clockwise large 1`] = `
<g>
  <path
    d="M47.552825814757675,-15.450849718747369 A50,50 0 1 1 -47.552825814757675,-15.450849718747387L-42.79754323328191,-13.905764746872649 A45,45 0 1 0 42.79754323328191,-13.905764746872633 L47.552825814757675,-15.450849718747369"
    style={
      Object {
        "fill": "lime",
      }
    }
  />
</g>
`;

exports[`Arc renders correct clockwise small 1`] = `
<g>
  <path
    d="M47.552825814757675,-15.450849718747369 A50,50 0 0 1 3.061616997868383e-15,50L2.7554552980815448e-15,45 A45,45 0 0 0 42.79754323328191,-13.905764746872633 L47.552825814757675,-15.450849718747369"
    style={
      Object {
        "fill": "lime",
      }
    }
  />
</g>
`;
